### PR TITLE
feat: Add self-learning from PR review outcomes

### PR DESCRIPTION
**Pull request by haacked**
Add `/review-code learn` command to analyze what happened after code reviews and improve future reviews based on actual outcomes.

New commands:
- `learn <pr>` - Analyze outcomes of a specific PR
- `learn` - Batch analyze all unanalyzed merged PRs with reviews
- `learn --apply` - Apply accumulated learnings to context files

New scripts:
- parse-review-findings.sh - Extract structured findings from reviews
- learn-from-pr.sh - Cross-reference engine comparing Claude's findings with other reviewers and commit history
- learn-batch.sh - Find unanalyzed PRs with existing reviews
- learn-apply.sh - Synthesize learnings into context file updates

The system tracks four learning types:
- valid_catch: Claude found issue, code was fixed
- false_positive: Claude was wrong
- deferred: Valid issue, postponed
- missed_pattern: Other reviewer found, Claude missed

### CODE REVIEW

File: skills/review-code/scripts/parse-review-findings.sh [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    local finding_file=""
+    local finding_line=""
+    local finding_description=""
+
+    while IFS= read -r line || [[ -n "${line}" ]]; do
+        # Detect agent section headers (## Security Review, ## Performance Review, etc.)
+        if [[ "${line}" =~ ^##[[:space:]]+(Security|Performance|Correctness|Maintainability|Testing|Compatibility|Architecture|Frontend)[[:space:]]+Review ]]; then
+            # Save any pending finding before switching sections
+            if [[ "${in_finding}" == true ]] && [[ -n "${finding_file}" ]] && [[ -n "${finding_description}" ]]; then
+                finding_description=$(echo "${finding_description}" | sed 's/^[[:space:]]*//;s/[[:space:]]*$//' | head -c 500)
```
Comment (haacked) [inline skills/review-code/scripts/parse-review-findings.sh]: This 15-line jq command for saving findings is duplicated 7 times in this file. Consider extracting to a save_finding() helper function to improve maintainability.

*From: Maintainability Review (90% confidence)*

File: skills/review-code/scripts/parse-review-arg.sh [hunk 2]
```diff
@@ -36,6 +40,13 @@ if [[ "${arg}" == "find" ]]; then
     file_pattern="${remaining_args[2]:-}"
 fi
 
+# Check for 'learn' mode - if present, shift remaining args
+if [[ "${arg}" == "learn" ]]; then
```
Comment (haacked) [inline skills/review-code/scripts/parse-review-arg.sh]: This could conflict with a branch named 'learn' when using find mode. Consider adding a FIND_MODE guard to the condition.

*From: Compatibility Review (50% confidence)*

File: skills/review-code/scripts/learn-batch.sh [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    local days=30
+
+    # Parse arguments
+    while [[ $# -gt 0 ]]; do
+        case $1 in
+            --limit)
+                limit="$2"
+                shift 2
+                ;;
+            --days)
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-batch.sh:54]: Fixed in ac0dcb5 - Added validation that --limit and --days are positive integers with clear error messages.

File: skills/review-code/scripts/learn-batch.sh [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        # Extract PR number from filename (pr-123.md -> 123)
+        if [[ "${filename}" =~ ^pr-([0-9]+)\.md$ ]]; then
+            pr_number="${BASH_REMATCH[1]}"
+        else
+            continue
+        fi
+
+        # Check if already analyzed
+        local repo_key="${org}/${repo}"
+        if echo "${analyzed_data}" | jq -e --arg key "${repo_key}" --arg pr "${pr_number}" '.[$key][$pr] != null' > /dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-batch.sh]: Fixed in 143b20d - Added `// {}` fallback to handle missing repo keys gracefully instead of relying on error suppression.

File: skills/review-code/scripts/learn-apply.sh [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+main() {
+    local threshold=3
+
+    # Parse arguments
+    while [[ $# -gt 0 ]]; do
+        case $1 in
+            --threshold)
+                threshold="$2"
+                shift 2
+                ;;
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh:58]: Fixed in 1a183a9 - Added validation that --threshold is a positive integer with clear error message.

File: skills/review-code/scripts/learn-from-pr.sh [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    reviews=$(gh api "repos/${org}/${repo}/pulls/${pr_number}/reviews" --jq '
+        [.[] | select(.state != "PENDING") | {
+            id: .id,
+            author: .user.login,
+            state: .state,
+            body: .body,
+            submitted_at: .submitted_at
+        }]
+    ' 2> /dev/null || echo "[]")
+
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Fixed in 7518eed - Removed the unused reviews variable and unnecessary GitHub API call.

File: skills/review-code/scripts/learn-apply.sh [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    if [[ "${total_learnings}" -eq 0 ]]; then
+        echo '{"proposals":[],"summary":{"total_learnings":0,"grouped_patterns":0,"actionable_proposals":0}}'
+        exit 0
+    fi
+
+    # Group learnings by type and context
+    # Key = type + language + framework (if any)
+    local grouped
+    grouped=$(echo "${learnings}" | jq '
+        group_by(.type + "_" + (.context.language // "unknown") + "_" + (.context.framework // "none"))
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh]: Fixed in 5fc92c4 - Added sort_by before group_by to ensure proper grouping of learnings.

File: skills/review-code/scripts/learn-apply.sh [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+#       "grouped_patterns": 5,
+#       "actionable_proposals": 2
+#     }
+#   }
+
+set -euo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+LEARNINGS_DIR="${SCRIPT_DIR}/../learnings"
+CONTEXT_DIR="${SCRIPT_DIR}/../../../context"
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh]: Fixed in 1f33927 - Removed the unused CONTEXT_DIR variable.

File: skills/review-code/scripts/learn-orchestrator.sh [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            # Escape the value for JSON
+            value=$(echo "${value}" | jq -Rs '.')
+            json="${json},\"${key}\":${value}"
+        fi
+    done
+
+    json="${json}}"
+    echo "${json}"
+}
+
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-orchestrator.sh]: Fixed in 8e8df7d - Removed the unused output_json helper function.

File: skills/review-code/scripts/parse-review-arg.sh [hunk 10]
```diff
@@ -177,6 +198,47 @@ validate_draft_mode() {
...
+        return 0
+    elif [[ -z "${arg}" ]]; then
+        # Batch mode: analyze all unanalyzed PRs with reviews
+        build_json_output "learn" "learn_submode" "batch"
+        return 0
+    else
+        build_json_error "Invalid argument for learn mode: ${arg}. Use a PR number, PR URL, or no argument for batch mode."
+        exit 1
+    fi
+}
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/parse-review-arg.sh:241]: Tests for `detect_learn_mode` already exist in `tests/unit/test-parse-review-arg-functions.bats` (lines 533-641). These cover batch mode, single PR mode with number and URL, apply mode, invalid arguments, and mode collisions with find.

File: skills/review-code/scripts/parse-review-findings.sh [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    # Parse the review file and extract findings
+    local findings="[]"
+    local current_agent=""
+    local current_confidence=""
+    local in_finding=false
+    local finding_file=""
+    local finding_line=""
+    local finding_description=""
+
+    while IFS= read -r line || [[ -n "${line}" ]]; do
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/parse-review-findings.sh:90]: Added `tests/unit/test-parse-review-findings.bats` with 24 tests covering: agent section headers, all 4 file:line patterns, confidence markers, multi-line descriptions, edge cases, and JSON output validation.

File: skills/review-code/scripts/parse-review-findings.sh [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+            continue
+        fi
+
+        # Detect confidence markers in current context
+        if [[ "${line}" =~ \[([0-9]+)%\] ]] || [[ "${line}" =~ \(([0-9]+)%[[:space:]]*confidence\) ]]; then
+            current_confidence="${BASH_REMATCH[1]}"
+        fi
+
+        # Accumulate description lines when in a finding (skip empty lines and headers)
+        if [[ "${in_finding}" == true ]] && [[ -n "${line}" ]] && [[ ! "${line}" =~ ^# ]]; then
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/parse-review-findings.sh:180]: Covered in the same test file - see tests for Pattern 1-4, agent header detection, and confidence marker extraction.

File: skills/review-code/scripts/learn-from-pr.sh [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+            was_addressed="likely"
+        else
+            was_addressed="not_modified"
+        fi
+
+        claude_results=$(echo "${claude_results}" | jq --argjson finding "${finding_json}" \
+            --arg addressed "${was_addressed}" \
+            '. + [($finding + {addressed: $addressed})]')
+    done < <(echo "${claude_findings}" | jq -c '.[]')
+
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Added `tests/unit/test-learn-from-pr.bats` with 26 tests covering: argument parsing, PR identifier extraction, review file detection, cross-reference jq patterns, line proximity checks, date comparison, output structure, and prompts_needed logic.

File: skills/review-code/scripts/learn-from-pr.sh [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+        local was_addressed="unknown"
+        if echo "${files_changed_after_review}" | jq -e --arg f "${finding_file}" 'index($f) != null' > /dev/null 2>&1; then
+            was_addressed="likely"
+        else
+            was_addressed="not_modified"
+        fi
+
+        claude_results=$(echo "${claude_results}" | jq --argjson finding "${finding_json}" \
+            --arg addressed "${was_addressed}" \
+            '. + [($finding + {addressed: $addressed})]')
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Covered in the same test file - see tests for jq file matching, addressed/not_modified status calculation, and summary field computation.

File: skills/review-code/scripts/learn-batch.sh [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+            --arg merged_at "${pr_merged_at}" \
+            '. + [{
+                org: $org,
+                repo: $repo,
+                pr_number: ($pr_number | tonumber),
+                review_file: $review_file,
+                merged_at: $merged_at
+            }]')
+
+        count=$((count + 1))
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-batch.sh:224]: Added `tests/unit/test-learn-batch.bats` with 25 tests covering: argument validation, empty review root handling, PR file path parsing, analyzed.json jq patterns (including the // {} fix), and date cutoff calculations.

File: skills/review-code/scripts/learn-apply.sh [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    local actionable_proposals
+    actionable_proposals=$(echo "${proposals}" | jq 'length')
+
+    # Build final output
+    jq -n \
+        --argjson proposals "${proposals}" \
+        --argjson total "${total_learnings}" \
+        --argjson grouped "${grouped_patterns}" \
+        --argjson actionable "${actionable_proposals}" \
+        '{
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh:205]: Added `tests/unit/test-learn-apply.bats` with 23 tests covering: argument validation, empty learnings handling, grouping by type/language/framework, threshold logic, all learning types (false_positive, missed_pattern, valid_catch, deferred), target file determination, sort_by before group_by, and proposal structure.

File: skills/review-code/scripts/learn-from-pr.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Added `tests/unit/test-learn-from-pr.bats` with 26 tests covering: PR identifier parsing, review file detection, cross-reference logic (files_changed_after_review, addressed status), date comparison logic, and prompts_needed computation.

File: skills/review-code/scripts/learn-batch.sh (hunk 15 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-batch.sh:224]: Added `tests/unit/test-learn-batch.bats` with 25 tests covering: argument validation (--limit/--days as positive integers), path parsing for org/repo/PR extraction, analyzed.json jq patterns with missing key handling, and date cutoff calculations.

File: skills/review-code/scripts/learn-apply.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh:205]: Added `tests/unit/test-learn-apply.bats` with 23 tests covering: threshold logic, all 4 learning types (false_positive, missed_pattern, valid_catch, deferred), target file determination (language vs framework), grouping/sorting, and proposal content generation.

File: skills/review-code/scripts/learn-from-pr.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Covered by `tests/unit/test-learn-from-pr.bats` - see tests for jq file matching patterns, line proximity checks, and addressed status logic.

File: skills/review-code/scripts/parse-review-findings.sh [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        save_finding "${current_agent:-unknown}" "${current_confidence:-0}" "${finding_file}" "${finding_line}" "${finding_description}"
+    fi
+
+    echo "${findings}"
+}
+
+# Only run main if script is executed directly (not sourced)
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+    main "$@"
+fi
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/parse-review-findings.sh:202]: Added comprehensive test coverage in `tests/unit/test-parse-review-findings.bats` - 24 tests covering all patterns and edge cases.

File: skills/review-code/scripts/learn-from-pr.sh [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+            summary: {
+                claude_total: ($claude_findings | length),
+                claude_addressed: ([$claude_findings[] | select(.addressed == "likely")] | length),
+                claude_not_addressed: ([$claude_findings[] | select(.addressed == "not_modified")] | length),
+                other_total: ($other_findings | length),
+                other_caught_by_claude: ([$other_findings[] | select(.claude_caught == true)] | length),
+                other_missed_by_claude: ([$other_findings[] | select(.claude_caught == false)] | length)
+            }
+        }'
+}
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh:243]: Added comprehensive test coverage in `tests/unit/test-learn-from-pr.bats` - 26 tests covering PR identifier parsing, review file detection, cross-reference logic, and prompts_needed computation.

File: skills/review-code/scripts/learn-batch.sh [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+    done < <(find "${review_root}" -name "pr-*.md" -type f 2> /dev/null | sort -r)
+
+    echo "${unanalyzed_prs}"
+}
+
+# Only run main if script is executed directly (not sourced)
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+    main "$@"
+fi
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-batch.sh:238]: Added comprehensive test coverage in `tests/unit/test-learn-batch.bats` - 25 tests covering argument validation, path parsing, analyzed.json handling, and date cutoffs.

File: skills/review-code/scripts/learn-apply.sh [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+                grouped_patterns: $grouped,
+                actionable_proposals: $actionable
+            }
+        }'
+}
+
+# Only run main if script is executed directly (not sourced)
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+    main "$@"
+fi
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-apply.sh:218]: Added comprehensive test coverage in `tests/unit/test-learn-apply.bats` - 23 tests covering threshold logic, all learning types, target file determination, and proposal generation.

File: skills/review-code/scripts/learn-from-pr.sh [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+    # Get the list of files changed in commits after the review
+    local files_changed_after_review="[]"
+    while IFS= read -r commit_json; do
+        local commit_sha commit_date commit_epoch
+        commit_sha=$(echo "${commit_json}" | jq -r '.sha')
+        commit_date=$(echo "${commit_json}" | jq -r '.date')
+
+        # Convert commit date to epoch seconds for reliable comparison
+        if [[ "${OSTYPE}" == "darwin"* ]]; then
+            commit_epoch=$(date -j -f "%Y-%m-%dT%H:%M:%SZ" "${commit_date%+00:00}Z" +%s 2>/dev/null || date -j -f "%Y-%m-%dT%H:%M:%S%z" "${commit_date}" +%s 2>/dev/null || echo "0")
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Valid observation. The current implementation has two fallback paths that should handle both cases, but you're right that the `${commit_date%+00:00}Z` transformation is incorrect when the date already ends with Z. The first date format attempt will fail with the malformed timestamp, then it tries the %z format which also fails for Z suffix, and falls back to 0. This could cause commits to be incorrectly classified as before the review. Will address in a follow-up.

File: skills/review-code/scripts/helpers/config-helpers.sh [hunk 22]
```diff
@@ -92,3 +92,29 @@ load_config_safely() {
...
+        config_file="${HOME}/.claude/review-code.env"
+    fi
+
+    if [[ -n "${config_file}" ]]; then
+        load_config_safely "${config_file}"
+        review_root="${REVIEW_ROOT_PATH:-${HOME}/dev/ai/reviews}"
+    fi
+
+    echo "${review_root}"
+}
```
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/helpers/config-helpers.sh:120]: Good catch. The `get_review_root` helper is currently tested indirectly through the scripts that use it, but dedicated unit tests for config precedence (no config, each config path, invalid REVIEW_ROOT_PATH) would be valuable. Adding to the backlog for follow-up.

File: skills/review-code/scripts/learn-from-pr.sh (hunk 21 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/learn-from-pr.sh]: Fixed in bd2b2f2 - Now properly handles both Z suffix and offset formats by checking the suffix type first and normalizing +HH:MM to +HHMM for BSD date's %z format.

File: skills/review-code/scripts/helpers/config-helpers.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (haacked) [reply skills/review-code/scripts/helpers/config-helpers.sh:120]: Added in b03b321 - 6 new tests covering: default path when no config, primary config path, fallback config path, precedence rules, and handling of missing/empty REVIEW_ROOT_PATH.
